src/test/java/com/alphawallet/attestation/demo/AttestationValidityCheck.sol

```diff
@@ -0,0 +1,497 @@
+    bytes1 constant VISIBLE_STRING_TAG  = bytes1(0x1a); // decimal 26
+    bytes1 constant GENERAL_STRING_TAG  = bytes1(0x1b); // decimal 27
+    bytes1 constant UNIVERSAL_STRING_TAG = bytes1(0x1c); // decimal 28
+    bytes1 constant BMP_STRING_TAG      = bytes1(0x1e); // decimal 30
+    bytes1 constant UTF8_STRING_TAG     = bytes1(0x0c); // decimal 12
+
+    bytes1 constant CONSTRUCTED_TAG     = bytes1(0x20); // decimal 28
+
+    uint256 constant IA5_CODE = uint256(bytes32("IA5")); //tags for disambiguating content
+    uint256 constant DEROBJ_CODE = uint256(bytes32("OBJID"));
+
+    uint256 constant public fieldSize = 21888242871839275222246405745257275088696311157297823662689037894645226208583;
+    uint256 constant public curveOrder = 21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+    event Value(uint256 indexed val);
+    event RtnStr(bytes val);
+    event RtnS(string val);
+
+    uint256[2] private G = [ 12022136709705892117842496518378933837282529509560188557390124672992517127582,
+    6765325636686621066142015726326349598074684595222800743368698766652936798612 ];
+
+    uint256[2] private H = [ 12263903704889727924109846582336855803381529831687633314439453294155493615168,
+    1637819407897162978922461013726819811885734067940976901570219278871042378189 ];
+
+    bytes constant GPoint = abi.encodePacked(uint8(0x04), uint256(12022136709705892117842496518378933837282529509560188557390124672992517127582),
+        uint256(6765325636686621066142015726326349598074684595222800743368698766652936798612));
+
+    constructor() public
+    {
+        owner = msg.sender;
+    }
+
+    struct Length {
+        uint decodeIndex;
+        uint length;
+    }
+
+    struct Exponent {
+        bytes base;
+        bytes riddle;
+        bytes tPoint;
+        uint256 challenge;
+    }
+
+    struct OctetString {
+        uint decodeIndex;
+        bytes byteCode;
+    }
+
+    function decodeAttestation(bytes memory attestation) public view returns(bool)
+    {
+        Length memory len;
+        Exponent memory pok;
+        OctetString memory octet;
+
+        require(attestation[0] == (CONSTRUCTED_TAG | SEQUENCE_TAG));
+
+        len = decodeLength(attestation, 1);
+        octet.decodeIndex = len.decodeIndex;
+
+        //decode parts
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.base = octet.byteCode;
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.riddle = octet.byteCode;
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.challenge = bytesToUint(octet.byteCode);
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.tPoint = octet.byteCode;
+
+        //The hand optimised concat4 is more optimal than using abi.encodePacked (checked the gas costs of both)
+        bytes memory cArray = concat4(GPoint, pok.base, pok.riddle, pok.tPoint);
```

Comment:
The values of the two commitments used in the equality check should also be included in cArray